keps/sig-apps/2214-indexed-job/README.md

```diff
@@ -0,0 +1,566 @@
+```
+
+The Job controller doesn't add the environment variable if there is a name
+conflict with an existing environment variables in the Job's Pod template or if
+the user already specified another environment variable for the same annotation.
+
+### Job completion and restart policy
+
+When dealing with Indexed Jobs, the Job controller keeps track of Pod
+completions for each index from 0 to `.spec.completions - 1`.
+Once the controller notices that a Pod finishes successfully, it will not create
+another Pod with the same index.
+
+The Job controller considers a Job completed when there is at least one
+successful Pod for each completion index.
+
+If an entire Pod fails, such as when the Pod gets kicked off the node or if a
+container of the Pod fails and `.spec.template.spec.restartPolicy = "Never"`,
+the Job controller gets the completion index for the failed Pod and creates a
+new Pod with the same index. The application needs to handle restarts in a
+different Pod.
+
+The Pod might not be immediately replaced, as the number of active Pods could
+have hit the parallelism limit. Once another Pod finishes, the Job controller
+syncs the Job, scanning for unsatisfied indexes and creates the missing Pods for
+them.
+
+The kubelet handles container restarts as usual, according to the
+`spec.template.spec.restartPolicy`.
+
+<<[UNRESOLVED TBD Beta: Track completed indexes in Job status]>>
+Once [kubernetes/kubernetes#28486](https://github.com/kubernetes/kubernetes/issues/28486)
+is resolved:
+
+The Job controller keeps track of completed indexes in the Job status in a
+base64 encoded bitmask.
+The command `kubectl describe` displays the list of completed indexes in a
+human-readable compressed format. For example:
+
+```
+Completed Indexes: [1-25,28,30-32]
+``` 
+
+The command crops the list of indexes once it reaches 200 characters.
+
+```
+Completed Indexes: [1-25,28,30-32,...]
+``` 
+<<[/UNRESOLVED]>>
+
+### Job parallelism
+
+A user can change the number of active Pods for a Job changing `.spec.parallelism`
+(note that `.spec.completions` is an immutable field).
+
+When starting a Job or increasing the parallelism, the Job controller creates
+Pods with lower completion index first, as long as there is no other completed
+or running Pod with the same index. This is to make the controller behavior more
+predictable. We do not offer guarantees on creation order based on completion
+index.
+
+Scale down is unaffected by completion index.
+
+### Test Plan
+
+Unit, integration and E2E tests cover the following Indexed Job mechanics:
+
+  - Creation with indexed Pod names and index annotations.
+  - Scale up and down.
+  - Pod failures.
+  
+Additionally, we add unit tests for API defaulting and validation, with feature
+gate enabled and disabled.
+  
+### Graduation Criteria
+
+#### Alpha -> Beta Graduation
+
+- Complete features:
+  - Completion index in annotation
+  - Restart policy
```

Comment:
Shouldn't those features be part of Alpha?

Reply:
Correct, created a section for alpha.